server/src/main/java/io/druid/server/QueryResource.java

```diff
@@ -394,16 +384,56 @@ public void write(OutputStream outputStream) throws IOException, WebApplicationE
          .addData("peer", req.getRemoteAddr())
          .emit();
 
+      return context.gotError(e);
+    }
+    finally {
+      Thread.currentThread().setName(currThreadName);
+    }
+  }
+
+  protected class RequestContext
```

Comment:
@navis this is a good generic way to handle context, it is not related to this PR thought, can we make this a separate PR i guess it can be part of the common utils ?